libr/esil/esil_hooks.c

```diff
@@ -0,0 +1,363 @@
+		hooks->reg_read_observers && hooks->reg_write_observers)) {
+		r_id_storage_free (hooks->mem_read_observers);
+		r_id_storage_free (hooks->mem_write_observers);
+		r_id_storage_free (hooks->reg_read_observers);
+		r_id_storage_free (hooks->reg_write_observers);
+		free (hooks);
+		return NULL;
+	}
+	return hooks;
+}
+
+static bool free_hook_cb(void *user, void *data, ut32 id) {
+	free (data);
+	return true;
+}
+
+R_API void r_esil_hooks_free(REsilHooks *hooks) {
+	if (hooks) {
+		r_id_storage_foreach (hooks->mem_read_observers, free_hook_cb, NULL);
+		r_id_storage_foreach (hooks->mem_write_observers, free_hook_cb, NULL);
+		r_id_storage_foreach (hooks->reg_read_observers, free_hook_cb, NULL);
+		r_id_storage_foreach (hooks->reg_write_observers, free_hook_cb, NULL);
+	}
+	free (hooks);
+}
+
+R_API bool r_esil_set_mem_read_imp(REsil *esil, REsilImpHookMemReadCB imp, void *user) {
+	r_return_val_if_fail (esil && esil->hooks, false);
+	if (!esil->hooks->mem_read_implementation) {
+		esil->hooks->mem_read_implementation = R_NEW (REsilHook);
+	}
+	REsilHook *hook = esil->hooks->mem_read_implementation;
+	r_return_val_if_fail (hook, false);
+	hook->user = user;
+	hook->imr = imp;
+	return true;
+}
+
+R_API void r_esil_del_mem_read_imp(REsil *esil) {
+	r_return_if_fail (esil && esil->hooks);
+	R_FREE (esil->hooks->mem_read_implementation);
+}
+
+R_API bool r_esil_set_mem_write_imp(REsil *esil, REsilImpHookMemWriteCB imp, void *user) {
+	r_return_val_if_fail (esil && esil->hooks, false);
+	if (!esil->hooks->mem_write_implementation) {
+		esil->hooks->mem_write_implementation = R_NEW (REsilHook);
+	}
+	REsilHook *hook = esil->hooks->mem_write_implementation;
+	r_return_val_if_fail (hook, false);
+	hook->user = user;
+	hook->imw = imp;
+	return true;
+}
+
+R_API void r_esil_del_mem_write_imp(REsil *esil) {
+	r_return_if_fail (esil && esil->hooks);
+	R_FREE (esil->hooks->mem_write_implementation);
+}
+
+R_API bool r_esil_set_reg_read_imp(REsil *esil, REsilImpHookRegReadCB imp, void *user) {
+	r_return_val_if_fail (esil && esil->hooks, false);
+	if (!esil->hooks->reg_read_implementation) {
+		esil->hooks->reg_read_implementation = R_NEW (REsilHook);
+	}
+	REsilHook *hook = esil->hooks->reg_read_implementation;
+	r_return_val_if_fail (hook, false);
+	hook->user = user;
+	hook->irr = imp;
+	return true;
+}
+
+R_API void r_esil_del_reg_read_imp(REsil *esil) {
+	r_return_if_fail (esil && esil->hooks);
+	R_FREE (esil->hooks->reg_read_implementation);
+}
+
+R_API bool r_esil_set_reg_write_imp(REsil *esil, REsilImpHookRegWriteCB imp, void *user) {
+	r_return_val_if_fail (esil && esil->hooks, false);
+	if (!esil->hooks->reg_write_implementation) {
+		esil->hooks->reg_write_implementation = R_NEW (REsilHook);
+	}
+	REsilHook *hook = esil->hooks->reg_write_implementation;
+	r_return_val_if_fail (hook, false);
+	hook->user = user;
+	hook->irw = imp;
+	return true;
+}
+
+R_API void r_esil_del_reg_write_imp(REsil *esil) {
```

Comment:
i would prefer to have a smaller API  and less functions, we want to do the ADD, DEL, CLEAR operations on RR,RW,MM,MW array of callbacks. this array of pointers can be an RVec, so they are executed in the right order and hold a void*user pointer. im thinking about something like this:

`esil.hook(ADD, MR, cb, user);`

ADD = enum for the supported operations: ADD, DEL, CLEAR, ..
MR = memory read, just another enum with other stuff
cb = the user provided callback for the given esil operation hook
user = user data pointer associated with the given callback

The cb is used as a uniqueid for each event, this is, you can `esil.hook(DEL, cb, NULL);`

At the end this is very similar to what REvent, what are the limitations you can think about it?